opentelemetry/proto/trace/v1/trace.proto

```diff
@@ -43,29 +43,28 @@ message TracesData {
   repeated ResourceSpans resource_spans = 1;
 }
 
-// A collection of InstrumentationLibrarySpans from a Resource.
+// A collection of ScopeSpans from a Resource.
 message ResourceSpans {
   // The resource for the spans in this message.
   // If this field is not set then no resource info is known.
   opentelemetry.proto.resource.v1.Resource resource = 1;
 
-  // A list of InstrumentationLibrarySpans that originate from a resource.
-  repeated InstrumentationLibrarySpans instrumentation_library_spans = 2;
+  // A list of ScopeSpans that originate from a resource.
+  repeated ScopeSpans scope_spans = 2;
```

Comment:
Wouldn't this be a breaking change?  E.g. I couldn't send a message generated with the previous version to a receiver with this version, and the same if the versions were swapped.

Reply:
Yes, what @Aneurysm9 said.

It is not a breaking change for Protobuf binary wire format. It is a breaking change for JSON format but that is OK, JSON format is not stable.

I believe it is also OK to break generated code since we do not provide any guarantees for that. We probably should start providing some guarantees for that at some point, probably when we decide that JSON is stable, since these 2 are somewhat interconnected (field names are visible in JSON).

Reply:
I think it will have ripple effects on things that use that generated code.  For instance, when https://github.com/open-telemetry/opentelemetry-go/blob/main/exporters/otlp/otlptrace/clients.go#L51 is updated to a version with this change it will force that upgrade on any users providing their own proto objects to the client.

Reply:
@Aneurysm9 is that Client interface part of the API that is already promised to be stable?

Reply:
Yes.  It is [at v1.4.1](https://pkg.go.dev/go.opentelemetry.io/otel/exporters/otlp/otlptrace@v1.4.1#Client).  We were comfortable taking this dependency as it was our understanding that the OTLP data model was stable and only backward compatible changes would be made.  In hindsight, we probably should have insisted on a v1.0.0 release and clarification regarding stability for just the bits on-the-wire or for the generated code we provide for interop purposes.

Reply:
OK, so we have a problem here. Some possible approaches:
1. Don't make this change. The message names and field names are somewhat weird though. We can keep the names and explain that they really are a "scope" in the comments. I am not a big fan of this approach.
2. Otel Go SDK pays the price of leaking an unstable dependency in its stable interface. We break the dependency and the Go SDK interface breaks its Client interface for exporter implementors.
3. We make a major version increase in https://github.com/open-telemetry/opentelemetry-proto-go for this breaking change. Otel Go SDK implements a new ClientV2 interface and uses the new major version dependency in this interface. Both old and new Client interface will be supported in parallel.
4. Something else?

Reply:
+1 to not breaking JSON. The JSON exporter in JS is in fairly wide use and I think a break like this would have quite some impact on our users.

Reply:
FYI, JSON is explicitly marked "Alpha" in this repository and in the spec and we should be free to break it:
>"alpha" maturity APIs can change in incompatible ways at any time.

Reply:
I'm not sure what the right thing to do here is.  Based on the stability description in the README, it seems like we have an expectation that we should be able to change field names at any time, but only because JSON is unstable and once JSON is stable then changing field names would impact compatibility for users.  

I think the same needs to be extended to the generated identifiers for stable signals.  It doesn't seem right to say that traces are stable but we can still change the message format in ways that will break anything using code generated from it.

Reply:
>It doesn't seem right to say that traces are stable but we can still change the message format in ways that will break anything using code generated from it.

I tend to agree with you, but until now this was the thinking (that it is OK to break generated code). If we knew this was supposed to a limitation we would be a lot more careful with names and with declaring OTLP stable.

I do not think we can retroactively apply such stability limitation right now. If we don't want to break existing code then the possible approach is to honour semver and increase the major version number in the generated code (approach 3 above).

After this change we should consider having formal stability guarantees for generated code.

Reply:
So I believe `opentelemetry-go` made a mistake exposing publicly a type from a version `0.x` module and not sure if that is expected or normal. @tigrannajaryan opentelemetry-proto-go is at version `0.12.0` so it does not need a major version bump, because that will make it `1.0` :)

Reply:
@tigrannajaryan I think we can make the change to the name, but also say that in JSON receivers should still accept the old name. And we will do this in the collector.

Reply:
@bogdandrutu I understand that `opentelemetry-proto-go` doesn't need a major version bump itself because it is on 0.x. However with proposal number 3 in order for `opentelemetry-go` to avoid breaking their clients it needs to be able to consume 2 different versions of `opentelemetry-proto-go` simultaneously. This requires a major version bump or for `opentelemetry-proto-go` to expose the changed generated code using a different import path (which would be even more awkward).

I don't know if this is a good way, I am just trying to find a way to allow us to change the proto here and also at the same time avoid breaking `opentelemetry-go`.

Maybe there is another way.

Reply:
The other thing is to fix the `oteltrace` package. Deprecate that package with a fix dependency on the v0.12.0 and design a new package that does not expose this public.

Reply:
Will there ever be a point where the entirety of the API (including names) is considered stable?  I feel like this needs to be discussed because projects that internally generate code will have to change.  That seems to be the conventional definition of unstable.

The other option is to go through the process of updating a field in a backward and forward compatible way:

1. Add a new field, and deprecate the old.
2. Wait 12 months by the guidelines that we have set out.
3. Remove the old field.  (This is still technically a breaking change, but one that we have allowed ourselves)

Reply:
>The other option is to go through the process of updating a field in a backward and forward compatible way

This would mean we are applying the most strict version of stability guarantees. This seems excessive to me for something that was not explicitly promised to be stable and implicitly was marked to be unstable (one reason JSON is marked unstable is because we said we don't want to guarantee the current message/field names).

Reply:
>FYI, JSON is explicitly marked "Alpha" in this repository and in the spec and we should be free to break it:

While I agree technically, I would hope that we take these breaking changes seriously anyways in the interest of our users.

There are also other problems not completely related to JSON. If a field can always be renamed, the logic to transform an SDK object into an OTLP object can never be shared between two different exporters even when JSON is completely out of the picture (HTTP/proto and gRPC for example). Even if generated code is not used, the code that transforms the SDK object to the OTLP object needs to pick a field name to use in its output which is then used as input to the final format encoder (protobuf in this case).

Reply:
>implicitly was marked to be unstable (one reason JSON is marked unstable is because we said we don't want to guarantee the current message/field names).

I need to correct myself. The message/field names are actually explicitly declared unstable. This is written in the README of this repository:
>Note that maturity guarantees apply only to wire-level compatibility for the binary Protobuf serialization. Neither message, field, nor enum names of Protobuf messages are visible on the wire and are not considered part of the guarantees. We are free to make a change to the names.

Reply:
>the logic to transform an SDK object into an OTLP object can never be shared between two different exporters even when JSON is completely out of the picture (HTTP/proto and gRPC for example). Even if generated code is not used, the code that transforms the SDK object to the OTLP object needs to pick a field name to use in its output which is then used as input to the final format encoder (protobuf in this case).

I don't understand the problem that this describes. Please clarify.

Reply:
I suspect JSON is actually being used in production - changing the format IMO is acceptable, but it's a lot more tricky if it's a hard break. Ideally we have at least one version of the collector that supports both names in the OTLP receiver. Can the collector be updated to do a backwards-compatibility pass on the incoming JSON to allow it to support both names temporarily?

Reply:
To clarify what, I think, @dyladan is saying is the reason that go has this problem is because of shared code.  We have a tiered system of SDK type -> `otlptrace` (which marshals to protobuf) -> Client (which sends over the wire, HTTP or grpc).  The part that we can't share is the marshaling of SDK types into protobuf types because that would expose the protobufs as part of a public API.

Reply:
@MadVikingGod got it exactly. If I write code that transforms the SDK types into types that would be serializable to protobuf or JSON, I can't share that code between multiple exporter packages if field names might change. The only way to resolve that would be for the transformation code to return an object with both field names forever (and mark deprecated the old one). The JSON exporter would then be responsible for stripping all deprecated fields, also forever, in order to avoid bloating the payload size.

Additionally @anuraaga is correct JSON is for sure being used in production in browser environments.

Reply:
It also affects the types between the SDK and exporter. The interface for the exporter already has a type which uses the old name, and the SDK type already contains the old names. Should these types also be extended to have both names indefinitely?

Reply:
We're going to have to carry the [`instrumentation.Library` type](https://pkg.go.dev/go.opentelemetry.io/otel/sdk@v1.4.1/instrumentation#Library) indefinitely, and I expect we'll have to carry it in our SDK export structures alongside a new type for the instrumentation scope that duplicates the data.

Reply:
>Can the collector be updated to do a backwards-compatibility pass on the incoming JSON to allow it to support both names temporarily?

Probably. I am not sure how exactly we can do it since JSON unmarshaling is a Protobuf library function, but perhaps it is doable.

Reply:
FWIW, my idea is brute force - process an untyped JSON tree before passing it to the protobuf library, which should naturally be doable. The main concern could be performance, but we expect JSON to be slower than binary anyways so while the format is considered alpha, having a temporary performance hit for handling this compatibility issue seems OK. We generally except the burden of compatibility to be on a receiver rather than an exporter, though of course the other option is for the JSON exporters to populate the same data into both naming schemes, using a transformation on the untyped JSON on export for exporters using the generated protobuf code. For reference, since Java doesn't use generated protos, it would be relatively trivial with no significant performance penalty (other than transmitting the additional data) to export both structs if that were deemed the way forward.

Reply:
Let's assume that receivers (such as Collector) can gracefully handle the old and new names simultaneously in JSON payloads. I think it is doable one way or another.

@dyladan @Aneurysm9 @MadVikingGod The remaining question is are you able to handle this change in your SDKs gracefully? If so we can move forward and merge the PR and you can upgrade to the new proto version when you are ready.

We can also bring this up in the maintainers meeting to make sure other language maintainers also have a chance to review.

Reply:
> The other option is to go through the process of updating a field in a backward and forward compatible way:
> 
> 1. Add a new field, and deprecate the old.
> 2. Wait 12 months by the guidelines that we have set out.
> 3. Remove the old field.  (This is still technically a breaking change, but one that we have allowed ourselves)

If we decide that we do want to gracefully handle old and new names in JSON receivers the above option may be the easiest from implementation perspective. (We can discuss whether it needs to be 12 months or less of grace period).

Reply:
> Let's assume that receivers (such as Collector) can gracefully handle the old and new names simultaneously in JSON payloads. I think it is doable one way or another.

This assumes our customers can update their collectors. If the SDK is updated with a minor version number, the user could reasonably expect it would continue to work with the collector version they're already using.

> @dyladan @Aneurysm9 @MadVikingGod The remaining question is are you able to handle this change in your SDKs gracefully? If so we can move forward and merge the PR and you can upgrade to the new proto version when you are ready.

In the SDK type we can simply duplicate the field with a new name and deprecate the old name. It is a small overhead, but I think not that big of a deal as it is just local and not over the wire.

Reply:
> This assumes our customers can update their collectors. If the SDK is updated with a minor version number, the user could reasonably expect it would continue to work with the collector version they're already using.

If we want to be extra careful for some grace period the senders can also double-publish the data under both names in JSON.

Reply:
>If we want to be extra careful for some grace period the senders can also double-publish the data under both names in JSON

That is what we will likely have to do. The field will have to be duplicated essentially everywhere.

- SDK export pipeline
- JSON serialization
- SDK -> OTLP transformer